stdlib/obj.ml

```diff
@@ -196,3 +196,128 @@ module Ephemeron = struct
+    external runstack : ('a, 'b) stack -> ('c -> 'a) -> 'c -> 'b = "%runstack"
+
+
+
+  module Deep = struct
+
+    type ('a,'b) continuation
+    type last_fiber
+
+    external take_cont_noexc : ('a, 'b) continuation -> ('a, 'b) stack =
+      "caml_continuation_use_noexc" [@@noalloc]
+    external alloc_stack :
+      ('a -> 'b) ->
+      (exn -> 'b) ->
+      ('c eff -> ('c, 'b) continuation -> last_fiber -> 'b) ->
+      ('a, 'b) stack = "caml_alloc_stack"
+
+    let continue k v = resume (take_cont_noexc k) (fun x -> x) v
+
+    let discontinue k e = resume (take_cont_noexc k) (fun e -> raise e) e
+
+    type ('a,'b) handler =
+      { retc: 'a -> 'b;
+        exnc: exn -> 'b;
+        effc: 'c.'c eff -> (('c,'b) continuation -> 'b) option }
+
+    external reperform :
+      'a eff -> ('a, 'b) continuation -> last_fiber -> 'b = "%reperform"
+
+    let match_with comp arg handler =
+      let effc eff k last_fiber =
+        match handler.effc eff with
+        | Some f -> f k
+        | None -> reperform eff k last_fiber
+      in
+      let s = alloc_stack handler.retc handler.exnc effc in
+      runstack s comp arg
+
+    type 'a effect_handler =
+      { effc: 'b. 'b eff -> (('b,'a) continuation -> 'a) option }
+
+    let try_with comp arg handler =
+      let effc' eff k last_fiber =
+        match handler.effc eff with
+        | Some f -> f k
+        | None -> reperform eff k last_fiber
+      in
+      let s = alloc_stack (fun x -> x) (fun e -> raise e) effc' in
+      runstack s comp arg
+
+    external drop_continuation : ('a,'b) continuation -> unit =
+      "caml_drop_continuation"
+  end
+
+  module Shallow = struct
+
+    type ('a,'b) continuation
+    type last_fiber
+
+    external alloc_stack :
+      ('a -> 'b) ->
+      (exn -> 'b) ->
+      ('c eff -> ('c, 'b) continuation -> last_fiber -> 'b) ->
+      ('a, 'b) stack = "caml_alloc_stack"
+
+
+    let fiber : type a b. (a -> b) -> (a, b) continuation = fun f ->
+      let module M = struct type _ eff += Initial_setup__ : a eff end in
+      let exception E of (a,b) continuation in
+      let f' () = f (perform M.Initial_setup__) in
+      let error _ = failwith "impossible" in
+      let effc eff k _last_fiber =
+        match eff with
+        | M.Initial_setup__ -> raise (E k)
+        | _ -> error ()
+      in
+      let s = alloc_stack error error effc in
+      try magic (runstack s f' ()) with E k -> k
+
+    type ('a,'b) handler =
+      { retc: 'a -> 'b;
+        exnc: exn -> 'b;
+        effc: 'c.'c eff -> (('c,'a) continuation -> 'b) option }
+
+    external update_handler :
+      ('a,'b) continuation ->
+      ('b -> 'c) ->
+      (exn -> 'c) ->
+      ('d eff -> ('d,'b) continuation -> last_fiber -> 'c) ->
+      ('a,'c) stack = "caml_continuation_use_and_update_handler_noexc" [@@noalloc]
+
+    external reperform :
+      'a eff -> ('a, 'b) continuation -> last_fiber -> 'c = "%reperform"
+
+    let continue_with k v handler =
```

Comment:
For what it is worth, this isn't exactly shallow handlers as they are presented in the literature. As by this interface you are obliged to install a handler upon resume, which has the nice property that when a subsequent effect operation is invoked there will be a handler in place. However, genuine shallow continuations need to not be run under a new handler, rather they can run in the ambient environment as is. Obviously, one can "simulate" this behaviour by installing the identity handler, though, such a simulation is not space-efficient.

Reply:
(This interface is the same as the interface we have in the "typed continuations" proposal for Wasm. I don't think it has a proper name in the literature yet.)

Reply:
I agree that this is not the standard presentation in literature but matches the typed continuations proposal for Wasm. I think the reasoning towards going for this interface is exactly what you have described, which is that there is always a handler guaranteed to be in scope.